### PR TITLE
XW-4372: Test templates in RTE

### PR DESCRIPTION
**Pull request by TK-999**
https://wikia-inc.atlassian.net/browse/XW-4372

### CODE REVIEW

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java [hunk 1]
```diff
@@ -148,6 +153,18 @@ public void addContent(String content) {
     PageObjectLogging.log("addContent", "content " + content + " added to the article", true);
   }
 
+  public boolean checkPortableInfoboxVisible() {
+    try {
+      wait.forElementVisible(iframe);
+      verifyComponent(portableInfoboxTransclusion);
+      return true;
+    } catch (TimeoutException e) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: MAJOR Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java [hunk 2]
```diff
@@ -319,6 +335,26 @@ public void verifyComponentRemoved(Components component) {
...
+    removeInfoboxButton.click();
+    removeConfirmationButton.click();
+  }
+
+  public boolean checkPortableInfoboxIsNotPresent() {
+    driver.switchTo().frame(iframe);
+    try {
+      wait.forElementNotPresent(portableInfoboxBy);
+      return true;
+    } catch (TimeoutException e) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: MAJOR Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [VisualEditModePageObject.java#L174](https://github.com/Wikia/selenium-tests/blob/19ebe54993aa6673d3606253eaeae367164b4c62/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java#L174): Remove this unused private "verifyComponent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [VisualEditModePageObject.java#L19](https://github.com/Wikia/selenium-tests/blob/19ebe54993aa6673d3606253eaeae367164b4c62/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java#L19): Remove this unused import 'org.openqa.selenium.JavascriptExecutor'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [VisualEditModePageObject.java#L23](https://github.com/Wikia/selenium-tests/blob/19ebe54993aa6673d3606253eaeae367164b4c62/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java#L23): Remove this unused import 'org.openqa.selenium.WebDriver'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [VisualEditModePageObject.java#L168](https://github.com/Wikia/selenium-tests/blob/12fd7f01db3797ff91d6542957357a81f5853def/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java#L168): Remove this unused private "verifyComponent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java [hunk 3]
```diff
@@ -289,144 +258,55 @@ public void removeComponent(Components component) {
...
-    WebDriverWait wait = new WebDriverWait(driver, ELEMENT_SHOW_UP_TIMEOUT);
-    wait.until(new ExpectedCondition<Boolean>() {
-      public Boolean apply(WebDriver driver) {
-        return categoryList.size() > categoryListSizeBeforeSubmission ? true : false;
+  public boolean checkPortableInfoboxIsNotPresent() {
+    return editorFrame.frameScope(() -> {
+      try {
+        wait.forElementNotPresent(portableInfoboxBy);
+        return true;
+      } catch (TimeoutException e) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: MAJOR Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java [hunk 4]
```diff
@@ -289,144 +258,55 @@ public void removeComponent(Components component) {
...
-    JavascriptExecutor js = (JavascriptExecutor) driver;
-    String returned = (String) js.executeScript("return $('ul.ui-autocomplete li').text()");
-    while (returned.isEmpty() && timeout <= 5000) {
-      try {
-        Thread.sleep(500);
-        timeout += 500;
-      } catch (InterruptedException e) {
-        PageObjectLogging
-            .log("triggerCategorySuggestions", "Interrupted Exception occurred", false);
+    editorFrame.frameScope(() -> {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: MAJOR Reduce this lambda expression number of lines from 22 to at most 20. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1188)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java [hunk 5]
```diff
@@ -131,33 +116,27 @@ public AceEditor getAceEditor() {
...
+  }
+
   private void verifyComponent(WebElement component) {
-    driver.switchTo().frame(iframe);
-    wait.forElementVisible(component);
-    driver.switchTo().defaultContent();
+    editorFrame.frameScope(() -> wait.forElementVisible(component));
   }
 
   private void verifyComponent(By componentBy) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: MAJOR Remove this unused private "verifyComponent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/ff438888476219d4785f3dab941d7fbe011889ed/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/ff438888476219d4785f3dab941d7fbe011889ed/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/32e59a3753642f065e7a9e7473f2cf6339142780/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/32e59a3753642f065e7a9e7473f2cf6339142780/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/10ec2f5c5dc484384d807c72e75d1d9dca365f0d/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/10ec2f5c5dc484384d807c72e75d1d9dca365f0d/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/04bc04de2c7f51396fe9dd624644ff0e63cbe4b2/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/04bc04de2c7f51396fe9dd624644ff0e63cbe4b2/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java (hunk 5 above)
Comment (rybmat) [reply src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: @TK-999 ☝️

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/rte/CategoryModule.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+   * Activate category suggestions based on the current content of the new category input field.
+   * The first item will be selected and added as a new category.
+   *
+   * @return the name of the newly added category from suggestions
+   */
+  public String selectFirstSuggestedCategory() {
+    WebElement suggestion = categorySelectModule.findElement(autoCompleteSuggestion);
+    String categoryName = suggestion.getText().trim();
+
+    suggestion.click();
```
Comment (rybmat) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/rte/CategoryModule.java]: `wait.forElementClickable()` should be called before click

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/rte/FeaturesModule.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+  private final WebDriver webDriver;
+
+  public FeaturesModule(WebDriver webDriver) {
+    PageFactory.initElements(webDriver, this);
+    this.webDriver = webDriver;
+  }
+
+  public InfoboxChoiceModal openInfoboxChoiceModal() {
+    infoboxButton.click();
```
Comment (rybmat) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/rte/FeaturesModule.java]: wait.forElementClickable

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java (hunk 5 above)
Comment (TK-999) [reply src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/VisualEditModePageObject.java]: sonar false positive

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/6f5f20f0dc08ffe49c79ae97b6ae7ea68da245b6/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/6f5f20f0dc08ffe49c79ae97b6ae7ea68da245b6/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/ff771f129f0bb19e32cbdefbb1237a1ea1f00118/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/ff771f129f0bb19e32cbdefbb1237a1ea1f00118/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/131ed19e42bfcd0cd62a2a24c04353f58a72ceb2/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/131ed19e42bfcd0cd62a2a24c04353f58a72ceb2/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/593a6abf7d67cd165372899d341e3ad2b4f7c0b5/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/593a6abf7d67cd165372899d341e3ad2b4f7c0b5/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/02c02521cbea97cccd980e2350b8b2b232731937/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/02c02521cbea97cccd980e2350b8b2b232731937/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 5 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L79](https://github.com/Wikia/selenium-tests/blob/5944d2b01a5ee1a8eeab13bc26fcf33e5cabdc81/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L79): Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS00117)
1. ![MAJOR][MAJOR] [SourceEditModePageObject.java#L324](https://github.com/Wikia/selenium-tests/blob/5944d2b01a5ee1a8eeab13bc26fcf33e5cabdc81/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/article/editmode/SourceEditModePageObject.java#L324): Remove this unused private "appendContent" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**